File: server-security/validators.ts
/**
 * @file validators.ts
 * @description Advanced security validators for validating user input, Unicode normalization,
 * control character filtering, size restriction, and prompt injection mitigation.
 */

// List of signatures commonly associated with prompt injection attacks
const PROMPT_INJECTION_SIGNATURES = [
  "ignore previous instructions",
  "ignore above instructions",
  "ignore the instructions",
  "reveal system prompt",
  "reveal your system prompt",
  "show system prompt",
  "expose system prompt",
  "what is your system prompt",
  "forget your rules",
  "forget previous instructions",
  "forget above rules",
  "execute javascript",
  "execute command",
  "show api key",
  "reveal api key",
  "expose api key",
  "developer instructions",
  "system prompt leak",
  "return environment variables",
  "display .env",
  "cat .env",
  "sudo rm",
  "eval(",
  "process.env"
];

/**
 * Validates, cleans, and normalizes user message inputs.
 * Returns the normalized message if valid, or throws a security error if validation fails.
 */
export function validateAndCleanMessage(message: any): string {
  if (message === undefined || message === null) {
    throw new Error("Message parameter is missing.");
  }

  if (typeof message !== "string") {
    throw new Error("Message must be a string value.");
  }

  // 1. Unicode Normalization (Form C)
  let cleanMessage = message.normalize("NFC");

  // 2. Reject excessively large payloads (DoS prevention)
  if (cleanMessage.length > 2000) {
    throw new Error("Message length exceeds maximum safe size of 2000 characters.");
  }

  // 3. Strip hidden / zero-width spaces / control characters
  // Allows standard newlines (\n, \r) and tabs (\t), rejects other control characters
  cleanMessage = cleanMessage.replace(/[\u200B-\u200D\uFEFF]/g, "");
  
  // Control characters regex (excluding whitespace characters)
  const controlCharsRegex = /[\x00-\x08\x0B\x0C\x0E-\x1F\x7F-\x9F]/g;
  if (controlCharsRegex.test(cleanMessage)) {
    throw new Error("Unsafe control characters detected in the request.");
  }

  // 4. Trim extra leading/trailing whitespace
  cleanMessage = cleanMessage.trim();

  if (!cleanMessage) {
    throw new Error("Message cannot be blank or contain only empty spacing.");
  }

  return cleanMessage;
}

/**
 * Evaluates the input message for potential prompt injection patterns.
 * Returns true if an injection signature is detected, false otherwise.
 */
export function detectPromptInjection(message: string): boolean {
  const lowercaseMsg = message.toLowerCase();
  
  for (const signature of PROMPT_INJECTION_SIGNATURES) {
    if (lowercaseMsg.includes(signature)) {
      return true;
    }
  }
  
  return false;
}

File: utils/validators.py
# -*- coding: utf-8 -*-
"""
@file validators.py
@description Performs string level input constraints, Unicode normalization,
and prompt injection detection for the Python Flask backend.
"""

import re
import unicodedata

# Common prompt injection signatures to scan for
PROMPT_INJECTION_SIGNATURES = [
    "ignore previous instructions",
    "ignore above instructions",
    "ignore the instructions",
    "reveal system prompt",
    "reveal your system prompt",
    "show system prompt",
    "expose system prompt",
    "what is your system prompt",
    "forget your rules",
    "forget previous instructions",
    "forget above rules",
    "execute javascript",
    "execute command",
    "show api key",
    "reveal api key",
    "expose api key",
    "developer instructions",
    "system prompt leak",
    "return environment variables",
    "display .env",
    "cat .env",
    "sudo rm",
    "eval(",
    "process.env"
]

def validate_and_normalize_message(message: str) -> str:
    """
    Validates the text parameter, normalizes Unicode encoding, limits size, 
    and filters invisible control characters.
    """
    if message is None:
        raise ValueError("Message input is missing.")
        
    if not isinstance(message, str):
        raise TypeError("Message input must be a string value.")

    # 1. Unicode Normalization (NFC)
    normalized = unicodedata.normalize("NFC", message)

    # 2. Size limit checking (DoS prevention)
    if len(normalized) > 2000:
        raise ValueError("Message length exceeds maximum safe size of 2000 characters.")

    # 3. Hidden control characters check (excluding tab, space, newline)
    # Zero-width spaces, etc.
    normalized = re.sub(r'[\u200b-\u200d\ufeff]', '', normalized)

    # Unicode control character categories (Cc, Cf)
    control_chars = [c for c in normalized if unicodedata.category(c) in ("Cc", "Cf") and c not in ("\n", "\r", "\t")]
    if control_chars:
        raise ValueError("Unsafe control characters detected in input.")

    # 4. Strip excessive spacing
    normalized = normalized.strip()

    if not normalized:
        raise ValueError("Message cannot be empty or containing only whitespaces.")

    return normalized

def detect_prompt_injection(message: str) -> bool:
    """
    Scans the string for common prompt injection patterns.
    """
    lowercase_msg = message.lower()
    for signature in PROMPT_INJECTION_SIGNATURES:
        if signature in lowercase_msg:
            return True
    return False
